transport/internal/tlsmux/listener_test.go

```diff
@@ -148,6 +203,103 @@ func TestMux(t *testing.T) {
 			n, err := conn.Read(response)
 			assert.NoError(t, err, "unexpected error")
 			assert.Equal(t, tt.body, response[:n], "unexpected response")
+			assert.Contains(t, root.Snapshot().Counters, tt.expectedCounter, "unexpected counters")
 		})
 	}
 }
+
+func TestConcurrentConnections(t *testing.T) {
+	defer goleak.VerifyNone(t)
+
+	scenario := tlsscenario.Create(t, time.Minute, time.Minute)
+	serverTlsConfig := &tls.Config{
+		GetCertificate: func(_ *tls.ClientHelloInfo) (*tls.Certificate, error) {
+			return &tls.Certificate{
+				Certificate: [][]byte{scenario.ServerCert.Raw},
+				Leaf:        scenario.ServerCert,
+				PrivateKey:  scenario.ServerKey,
+			}, nil
+		},
+		MinVersion: tls.VersionTLS13,
+		ClientAuth: tls.RequireAndVerifyClientCert,
+		ClientCAs:  scenario.CAs,
+	}
+	clientTlsConfig := &tls.Config{
+		GetClientCertificate: func(_ *tls.CertificateRequestInfo) (*tls.Certificate, error) {
+			return &tls.Certificate{
+				Certificate: [][]byte{scenario.ClientCert.Raw},
+				Leaf:        scenario.ClientCert,
+				PrivateKey:  scenario.ClientKey,
+			}, nil
+		},
+		MinVersion: tls.VersionTLS10,
+		MaxVersion: tls.VersionTLS13,
+		RootCAs:    scenario.CAs,
+	}
+
+	lis, err := net.Listen("tcp", "127.0.0.1:0")
+	require.NoError(t, err, "unexpected error on listening")
+
+	muxLis := tlsmux.NewListener(tlsmux.Config{
+		Listener:      lis,
+		TLSConfig:     serverTlsConfig,
+		Meter:         metrics.New().Scope(),
+		Logger:        zap.NewNop(),
+		ServiceName:   "test-svc",
+		TransportName: "test-transport",
+	})
+	defer muxLis.Close()
+
+	msg := "hello world"
+	totalConnections := 100
+
+	var wg sync.WaitGroup
+	defer wg.Wait()
+	for i := 0; i < totalConnections; i++ {
+		wg.Add(1)
+		go func(id int) {
+			defer wg.Done()
+			var conn net.Conn
+			var err error
+			if id%2 == 0 {
+				conn, err = tls.Dial(lis.Addr().Network(), lis.Addr().String(), clientTlsConfig)
+			} else {
+				conn, err = net.Dial(lis.Addr().Network(), lis.Addr().String())
+			}
+
+			require.NoError(t, err, "unexpected error on dial")
+			defer conn.Close()
+
+			n, err := conn.Write([]byte(msg))
+			require.NoError(t, err, "unexpected error on client write")
+			assert.Equal(t, len(msg), n, "unexpected write length")
+
+			buf := make([]byte, len(msg))
+			n, err = io.ReadFull(conn, buf)
+			require.NoError(t, err)
+			assert.Equal(t, len(msg), n)
+			assert.Equal(t, msg, string(buf))
+		}(i)
+	}
+
+	for i := 0; i < totalConnections; i++ {
+		conn, err := muxLis.Accept()
+		require.NoError(t, err)
+
+		wg.Add(1)
+		go func(c net.Conn) {
+			defer wg.Done()
+
+			buf := make([]byte, len(msg))
```

Comment:
looks like both tls and plain client are writing same message. Should we write different msg to both connection to ensure there are being processed by right set of routines?

Reply:
Sorry, this wasn't rebased during the time of review. I'll address this in a separate pull request